File: projects/iffywow/node/src/wire.js
/**
 * Wire form — JSON tagged arrays (order-preserving; never free-form objects).
 * Mirrors `web/src/coordinate.js` (same acceptance/rejection behavior, with
 * registry error codes) and CODEC.md §2:
 *
 *   word      := ["ithkuil-word", version, [glyph, ...]]
 *   glyph     := ["glyph", character_class, base, orientation, [socket, ...]]
 *   socket    := [socket_id, null | modifier]
 *   modifier  := ["modifier", shape, orientation, [diacritic_id, ...], [socket, ...]]
 *
 * `fromWire` is LENIENT-canonicalizing: unsorted sockets are sorted, empty
 * sockets (`[id, null]`) are accepted on input and dropped, duplicate socket
 * ids are always an error. `toWire` emits the canonical form: sockets
 * sorted, empties absent — one meaning, one representation.
 */

import { IthkuilError, show } from "./errors.js";
import { validate, canonicalize, MAX_SOCKET_ID, MAX_ORIENTATION } from "./coord.js";

export const WORD_TAG = "ithkuil-word";
export const GLYPH_TAG = "glyph";
export const MODIFIER_TAG = "modifier";

function fail(code, msg, path) {
  throw new IthkuilError(code, `invalid coordinate at ${path}: ${msg}`);
}

function nat(v, path) {
  if (typeof v !== "number" || !Number.isSafeInteger(v) || v < 0) {
    fail("invalid_structure", "expected non-negative integer", path);
  }
  return v === 0 ? 0 : v; // normalize -0 -> 0
}

function orient(v, path) {
  const n = nat(v, path);
  if (n > MAX_ORIENTATION) fail("invalid_orientation", `orientation must be 0..${MAX_ORIENTATION}`, path);
  return n;
}

/**
 * Parse a wire coordinate (tagged arrays, or a JSON text containing them)
 * into a canonical, deep-frozen Coord.
 */
export function fromWire(input) {
  let value = input;
  if (typeof value === "string") {
    value = parseJson(value); // single level of JSON decoding, as in web/
  }
  return parseWireValue(value);
}

/** Canonical Coord (or lenient native coord) -> canonical wire tagged arrays. */
export function toWire(coord) {
  const word = canonicalize(coord);
  return [WORD_TAG, word.version, word.glyphs.map(glyphWire)];
}

/** Canonical wire JSON text. Matches `JSON.stringify` of `toWire(coord)`. */
export function toWireJson(coord) {
  return JSON.stringify(toWire(coord));
}

/** Parse wire JSON text into a canonical, deep-frozen Coord. */
export function fromWireJson(text) {
  if (typeof text !== "string") {
    throw new IthkuilError("invalid_structure", `expected wire JSON text, got ${show(text)}`);
  }
  return parseWireValue(parseJson(text));
}

// --------------------------------------------------------------------------
// Internals.
// --------------------------------------------------------------------------

function parseJson(text) {
  try {
    return JSON.parse(text);
  } catch (e) {
    throw new IthkuilError("invalid_structure", `invalid coordinate: not JSON (${e.message})`);
  }
}

function parseWireValue(value) {
  if (!Array.isArray(value) || value.length !== 3 || value[0] !== WORD_TAG) {
    fail("invalid_structure", `expected ["${WORD_TAG}", version, glyphs]`, "$");
  }
  const version = value[1];
  if (typeof version !== "number" || !Number.isSafeInteger(version) || version < 1) {
    fail("invalid_version", "schema version must be a positive integer", "$[1]");
  }
  if (!Array.isArray(value[2])) fail("invalid_structure", "glyph list must be an array", "$[2]");
  const glyphs = value[2].map((g, i) => parseGlyph(g, `$[2][${i}]`));
  return validate({ version, glyphs }); // already canonical here; validate deep-freezes
}

function parseGlyph(g, path) {
  if (!Array.isArray(g) || g.length !== 5 || g[0] !== GLYPH_TAG) {
    fail("invalid_structure", `expected ["${GLYPH_TAG}", class, base, orientation, sockets]`, path);
  }
  return {
    characterClass: nat(g[1], `${path}[1]`),
    base: nat(g[2], `${path}[2]`),
    orientation: orient(g[3], `${path}[3]`),
    sockets: parseSockets(g[4], `${path}[4]`),
  };
}

function parseSockets(list, path) {
  if (!Array.isArray(list)) fail("invalid_structure", "sockets must be an array", path);
  const seen = new Set();
  const parsed = list.map((s, i) => {
    const p = `${path}[${i}]`;
    if (!Array.isArray(s) || s.length !== 2) fail("invalid_structure", "expected [socket_id, null | modifier]", p);
    const id = nat(s[0], `${p}[0]`);
    if (id > MAX_SOCKET_ID) {
      fail("invalid_socket_id", `socket id must be 0..${MAX_SOCKET_ID} in schema v1`, `${p}[0]`);
    }
    if (seen.has(id)) fail("duplicate_socket", `duplicate socket ${id}`, p); // even when one side is null
    seen.add(id);
    return [id, s[1] === null ? null : parseModifier(s[1], `${p}[1]`)];
  });
  parsed.sort((a, b) => a[0] - b[0]); // canonical ordering
  return parsed
    .filter(([, modifier]) => modifier !== null) // empties omitted
    .map(([id, modifier]) => ({ id, modifier }));
}

function parseModifier(m, path) {
  if (!Array.isArray(m) || m.length !== 5 || m[0] !== MODIFIER_TAG) {
    fail("invalid_structure", `expected ["${MODIFIER_TAG}", shape, orientation, diacritics, sockets]`, path);
  }
  if (!Array.isArray(m[3])) fail("invalid_structure", "diacritics must be an array", `${path}[3]`);
  return {
    shape: nat(m[1], `${path}[1]`),
    orientation: orient(m[2], `${path}[2]`),
    diacritics: m[3].map((d, i) => nat(d, `${path}[3][${i}]`)),
    sockets: parseSockets(m[4], `${path}[4]`),
  };
}

function glyphWire(g) {
  return [GLYPH_TAG, g.characterClass, g.base, g.orientation, g.sockets.map(socketWire)];
}

function socketWire(s) {
  return [s.id, modifierWire(s.modifier)];
}

function modifierWire(m) {
  return [MODIFIER_TAG, m.shape, m.orientation, [...m.diacritics], m.sockets.map(socketWire)];
}

File: projects/iffywow/rust/src/wire.rs
//! JSON wire form (tagged arrays). Normative: `../../CODEC.md` §2.
//!
//! ```text
//! ["ithkuil-word", version, [glyph, ...]]
//! ["glyph", character_class, base, orientation, [socket, ...]]
//! [socket_id, null | modifier]
//! ["modifier", shape, orientation, [diacritic_id, ...], [socket, ...]]
//! ```
//!
//! `to_wire` / `to_wire_json` are pure projections of a [`Word`] to its
//! tagged-array form. `from_wire` / `from_wire_json` are **lenient but
//! canonicalizing** per SDK-INTERFACE.md §2: unsorted sockets are sorted,
//! `[id, null]` empties are dropped, duplicated socket ids are always an
//! error (`duplicate_socket`).
//!
//! Numbers at this boundary: non-negative integers, including integral
//! floats (JSON cannot distinguish `2.0` from `2` — parity with the Python
//! reference). Values that cannot be represented exactly (above 2^53 on the
//! float path, above u64/u32 on the integer path) are rejected with
//! `invalid_structure`, never truncated.

use serde_json::Value;

use crate::coord::{Glyph, Modifier, Socket, Word, MAX_ORIENTATION, MAX_SOCKET_ID};
use crate::error::{Error, Result};

pub const WORD_TAG: &str = "ithkuil-word";
pub const GLYPH_TAG: &str = "glyph";
pub const MODIFIER_TAG: &str = "modifier";

// ---------------------------------------------------------------------------
// Word -> wire.
// ---------------------------------------------------------------------------

/// Canonical wire form (structural JSON value).
pub fn to_wire(word: &Word) -> Value {
    Value::Array(vec![
        Value::String(WORD_TAG.to_string()),
        Value::from(word.version),
        Value::Array(word.glyphs.iter().map(glyph_wire).collect()),
    ])
}

fn glyph_wire(glyph: &Glyph) -> Value {
    Value::Array(vec![
        Value::String(GLYPH_TAG.to_string()),
        Value::from(glyph.character_class),
        Value::from(glyph.base),
        Value::from(glyph.orientation),
        Value::Array(glyph.sockets.iter().map(socket_wire).collect()),
    ])
}

fn socket_wire(socket: &Socket) -> Value {
    Value::Array(vec![
        Value::from(socket.id),
        modifier_wire(&socket.modifier),
    ])
}

fn modifier_wire(modifier: &Modifier) -> Value {
    Value::Array(vec![
        Value::String(MODIFIER_TAG.to_string()),
        Value::from(modifier.shape),
        Value::from(modifier.orientation),
        Value::Array(modifier.diacritics.iter().map(|d| Value::from(*d)).collect()),
        Value::Array(modifier.sockets.iter().map(socket_wire).collect()),
    ])
}

/// Canonical wire form as compact JSON text (matches `JSON.stringify` /
/// Python `json.dumps(..., separators=(",", ":"))` of the same coordinate).
pub fn to_wire_json(word: &Word) -> String {
    serde_json::to_string(&to_wire(word))
        .expect("a tree of arrays, unsigned numbers and strings always serializes")
}

// ---------------------------------------------------------------------------
// Wire -> word (lenient, canonicalizing).
// ---------------------------------------------------------------------------

/// Parse a structural wire value into a canonical [`Word`].
pub fn from_wire(value: &Value) -> Result<Word> {
    let arr = match value.as_array() {
        Some(a) if a.len() == 3 && a[0].as_str() == Some(WORD_TAG) => a,
        _ => {
            return Err(fail(
                "$",
                &format!("expected [\"{WORD_TAG}\", version, glyphs]"),
            ))
        }
    };
    let version = match wire_nat(&arr[1], "$[1]") {
        Ok(v) if v >= 1 => v,
        _ => {
            return Err(Error::InvalidVersion(
                "invalid coordinate at $[1]: schema version must be a positive integer"
                    .to_string(),
            ))
        }
    };
    let version = u32::try_from(version).map_err(|_| {
        Error::InvalidStructure(
            "invalid coordinate at $[1]: word version exceeds the u32 range representable \
             by this SDK"
                .to_string(),
        )
    })?;
    let glyphs_arr = arr[2]
        .as_array()
        .ok_or_else(|| fail("$[2]", "glyph list must be an array"))?;
    let mut glyphs = Vec::with_capacity(glyphs_arr.len());
    for (i, glyph_value) in glyphs_arr.iter().enumerate() {
        glyphs.push(wire_glyph(glyph_value, &format!("$[2][{i}]"))?);
    }
    Ok(Word { version, glyphs })
}

/// Parse JSON text into a canonical [`Word`]. Text that is not valid JSON is
/// rejected with `invalid_structure`.
pub fn from_wire_json(text: &str) -> Result<Word> {
    let value: Value = serde_json::from_str(text).map_err(|e| {
        Error::InvalidStructure(format!("coordinate text is not valid JSON: {e}"))
    })?;
    from_wire(&value)
}

fn fail(path: &str, msg: &str) -> Error {
    Error::InvalidStructure(format!("invalid coordinate at {path}: {msg}"))
}

/// Non-negative integer at the JSON boundary, exact-value only.
fn wire_nat(value: &Value, path: &str) -> Result<u64> {
    if let Some(u) = value.as_u64() {
        return Ok(u);
    }
    if let Some(f) = value.as_f64() {
        // Integral floats are accepted (JSON `2.0` == `2`), but only inside
        // the exactly-representable range, so nothing is ever truncated.
        if f >= 0.0 && f.fract() == 0.0 && f <= 9_007_199_254_740_992.0 {
            return Ok(f as u64);
        }
    }
    Err(fail(path, "expected non-negative integer"))
}

fn wire_orientation(value: &Value, path: &str) -> Result<u8> {
    let n = wire_nat(value, path)?;
    if n > MAX_ORIENTATION as u64 {
        return Err(Error::InvalidOrientation(format!(
            "invalid coordinate at {path}: orientation must be 0..{MAX_ORIENTATION}"
        )));
    }
    Ok(n as u8)
}

fn wire_glyph(value: &Value, path: &str) -> Result<Glyph> {
    let arr = match value.as_array() {
        Some(a) if a.len() == 5 && a[0].as_str() == Some(GLYPH_TAG) => a,
        _ => {
            return Err(fail(
                path,
                &format!("expected [\"{GLYPH_TAG}\", class, base, orientation, sockets]"),
            ))
        }
    };
    Ok(Glyph {
        character_class: wire_nat(&arr[1], &format!("{path}[1]"))?,
        base: wire_nat(&arr[2], &format!("{path}[2]"))?,
        orientation: wire_orientation(&arr[3], &format!("{path}[3]"))?,
        sockets: wire_sockets(&arr[4], &format!("{path}[4]"))?,
    })
}

/// Lenient socket parsing: accepts any order and `[id, null]` empties; sorts
/// ascending and drops empties (canonical form). Duplicate ids error even
/// when one side is null.
fn wire_sockets(value: &Value, path: &str) -> Result<Vec<Socket>> {
    let arr = value
        .as_array()
        .ok_or_else(|| fail(path, "sockets must be an array"))?;
    let mut seen = [false; (MAX_SOCKET_ID as usize) + 1];
    let mut parsed: Vec<(u8, Option<Modifier>)> = Vec::with_capacity(arr.len());
    for (i, entry) in arr.iter().enumerate() {
        let entry_path = format!("{path}[{i}]");
        let pair = match entry.as_array() {
            Some(p) if p.len() == 2 => p,
            _ => return Err(fail(&entry_path, "expected [socket_id, null | modifier]")),
        };
        let id = wire_nat(&pair[0], &format!("{entry_path}[0]"))?;
        if id > MAX_SOCKET_ID as u64 {
            return Err(Error::InvalidSocketId(format!(
                "invalid coordinate at {entry_path}[0]: socket id must be \
                 0..{MAX_SOCKET_ID} in schema v1"
            )));
        }
        let id = id as u8;
        if seen[id as usize] {
            return Err(Error::DuplicateSocket(format!(
                "invalid coordinate at {entry_path}: duplicate socket {id}"
            )));
        }
        seen[id as usize] = true;
        let modifier = if pair[1].is_null() {
            None
        } else {
            Some(wire_modifier(&pair[1], &format!("{entry_path}[1]"))?)
        };
        parsed.push((id, modifier));
    }
    parsed.sort_by_key(|(id, _)| *id); // canonical ordering
    Ok(parsed
        .into_iter()
        .filter_map(|(id, modifier)| modifier.map(|modifier| Socket { id, modifier }))
        .collect()) // empties omitted
}

fn wire_modifier(value: &Value, path: &str) -> Result<Modifier> {
    let arr = match value.as_array() {
        Some(a) if a.len() == 5 && a[0].as_str() == Some(MODIFIER_TAG) => a,
        _ => {
            return Err(fail(
                path,
                &format!("expected [\"{MODIFIER_TAG}\", shape, orientation, diacritics, sockets]"),
            ))
        }
    };
    let diacritics_arr = arr[3]
        .as_array()
        .ok_or_else(|| fail(&format!("{path}[3]"), "diacritics must be an array"))?;
    let mut diacritics = Vec::with_capacity(diacritics_arr.len());
    for (i, d) in diacritics_arr.iter().enumerate() {
        diacritics.push(wire_nat(d, &format!("{path}[3][{i}]"))?);
    }
    Ok(Modifier {
        shape: wire_nat(&arr[1], &format!("{path}[1]"))?,
        orientation: wire_orientation(&arr[2], &format!("{path}[2]"))?,
        diacritics,
        sockets: wire_sockets(&arr[4], &format!("{path}[4]"))?,
    })
}
